x/clp/handler.go

```diff
@@ -0,0 +1,516 @@
+			sdk.NewAttribute(types.AttributeKeyPool, pool.String()),
+			sdk.NewAttribute(types.AttributeKeyHeight, strconv.FormatInt(ctx.BlockHeight(), 10)),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer.String()),
+		),
+	})
+	return &sdk.Result{Events: ctx.EventManager().Events()}, nil
+}
+
+func handleMsgCreatePool(ctx sdk.Context, keeper Keeper, msg MsgCreatePool) (*sdk.Result, error) {
+	// Verify min threshold
+	MinThreshold := keeper.GetParams(ctx).MinCreatePoolThreshold
+	if (msg.ExternalAssetAmount + msg.NativeAssetAmount) < MinThreshold { // Need to verify
+		return nil, types.ErrTotalAmountTooLow
+	}
+	// Check if pool already exists
+	if keeper.ExistsPool(ctx, msg.ExternalAsset.Ticker) {
+		return nil, types.ErrUnableToCreatePool
+	}
+
+	asset := msg.ExternalAsset
+	// Verify user has balance to create a new pool
+	externalAssetCoin := sdk.NewCoin(msg.ExternalAsset.Ticker, sdk.NewIntFromUint64(uint64(msg.ExternalAssetAmount)))
+	nativeAssetCoin := sdk.NewCoin(GetSettlementAsset().Ticker, sdk.NewIntFromUint64(uint64(msg.NativeAssetAmount)))
+	if !keeper.BankKeeper.HasCoins(ctx, msg.Signer, sdk.Coins{externalAssetCoin, nativeAssetCoin}) {
+		return nil, types.ErrBalanceNotAvailable
+	}
+
+	nativeBalance := msg.NativeAssetAmount
+	externalBalance := msg.ExternalAssetAmount
+	poolUnits, lpunits := calculatePoolUnits(0, 0, 0, nativeBalance, externalBalance)
+	pool, err := NewPool(asset, nativeBalance, externalBalance, poolUnits)
+	if err != nil {
+		return nil, errors.Wrap(types.ErrUnableToCreatePool, err.Error())
+	}
+	// Send coins from suer to pool
+	err = keeper.BankKeeper.SendCoins(ctx, msg.Signer, pool.PoolAddress, sdk.Coins{externalAssetCoin, nativeAssetCoin})
+	if err != nil {
+		return nil, err
+	}
+	// Pool creator becomes the first LP
+	lp := NewLiquidityProvider(asset, lpunits, msg.Signer)
+	err = keeper.SetPool(ctx, pool)
+	if err != nil {
+		return nil, errors.Wrap(types.ErrUnableToSetPool, err.Error())
+	}
+	keeper.SetLiquidityProvider(ctx, lp)
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeCreatePool,
+			sdk.NewAttribute(types.AttributeKeyPool, pool.String()),
+			sdk.NewAttribute(types.AttributeKeyHeight, strconv.FormatInt(ctx.BlockHeight(), 10)),
+		),
+		sdk.NewEvent(
+			types.EventTypeCreateLiquidityProvider,
+			sdk.NewAttribute(types.AttributeKeyLiquidityProvider, lp.String()),
+			sdk.NewAttribute(types.AttributeKeyHeight, strconv.FormatInt(ctx.BlockHeight(), 10)),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer.String()),
+		),
+	})
+	return &sdk.Result{Events: ctx.EventManager().Events()}, nil
+}
+
+func handleMsgAddLiquidity(ctx sdk.Context, keeper Keeper, msg MsgAddLiquidity) (*sdk.Result, error) {
```

Comment:
@musnit Yes, Thorchain's CLP model allows for both adding and removing liquidity asymmetrically.

Reply:
Does it not need to be done in the same way as with removing, with an asymmetry variable and a swap, as opposed to just adding as is?

Reply:
calculatePoolUnits() calculates ownership of the pool regardless of symmetry, there's a description here: https://medium.com/thorchain/asymmetric-withdrawals-on-bepswap-a6924ed2f28b.
Because we don't have any price tracking anywhere there wouldn't be a real way to add "symmetrically" to an empty pool.  The only symmetry we could have would be to add an equal number of tokens but that doesn't correlate to the tokens' value. The symmetry is determined by the market and arbitrage bots after the pools are created.

Reply:
Got it - yeh the medium article is a little sparse on info when it comes to details of asymmetric staking.

for asymmetric withdrawal the formula is effectively equivalent to a symmetric withdrawal and a swap, and our implementation does the latter.

does the same analogy not apply to asymmetric adding?

Reply:
In the case of withdrawals, our issue was that in order to implement asymmetry without a swap we would need to make up a new implementation of a formula.  But mathematically a symmetrical withdrawal and a swap is the same so it was simpler to go with symmetrical + swap.

For deposits the same logic is true: a swap to symmetry followed by a deposit would have the same slip impact as an asymmetrical deposit.  The difference is that we have a functional implementation from Thorchain of their ownership formula where we didn't have one for asymmetrical withdrawals. 

So it's the same situation in the opposite direction. To do a swap and symmetrical deposit we would need to add more steps and new logic.  We would also need to have a pricefeed module in order to determine what symmetry is for a new pool. 

There's a bit (but not much) more explanation in their docs: https://docs.thorchain.org/roles/liquidity-providers

> Asymmetrical deposits is where users deposit unequal values of 2 assets to a pool. For example, a user deposits $2000 of BTC and $0 of RUNE to the BTC/RUNE pool. Under the hood, the member is given an ownership of the pool that takes into account the slip created in the price. The liquidity provider will end up with <$1000 in BTC and <$1000 in RUNE. The deeper the pool, the closer to a total of $2000 the member will own.

>Note: there is no difference between swapping into symmetrical shares, then depositing that, or depositing asymmetrically and being arb'd to be symmetrical. You will still experience the same net slip.

Reply:
do LPs still get a fee when an asymmetric deposit is made?

Reply:
Yeah we don't currently have any system income set up but part of the model is that network fees are taken for withdrawals and lp rewards are paid through liquidity fees + network fee reserves.  

For asymmetric deposits the pool will become imbalanced and open up a swap arbitrage opportunity.  So a swap from an arbitrage bot will occur, securing the fee for lp's and resulting in proportionate loss of value for the depositer.

Reply:
yeh i was more thinking about a case in which someone makes an asymmetric deposit and then subsequent asymmetric withdrawal to avoid having to do any actual swap and so avoid paying fees

Reply:
The pool ownership formula takes into account the price slip created by the asymmetric deposit.  So the same value that the swap slip fee would grab is deducted from the depositer's ownership and therefore implicitly distributed to the rest of the pool's liquidity providers. It's not explicitly collected through a fee mechanism but through ownership calculation.

But the implementation of calculatePoolUnits that we grabbed doesn't factor in slip, this is an oversight on my part.  Thorchain has another implementation that accounts for slip that we need to use instead.

Reply:
@Jedi2002 we need to change calculatePoolUnits to reflect this implementation: 
```golang
// r = rune staked;
// a = asset staked
// R = rune Balance (before)
// A = asset Balance (before)
// P = existing Pool Units
// slipAdjustment = (1 - ABS((R a - r A)/((2 r + R) (a + A))))
// units = ((P (a R + A r))/(2 A R))*slidAdjustment
func calculatePoolUnitsV14(oldPoolUnits, poolRune, poolAsset, stakeRune, stakeAsset cosmos.Uint) (cosmos.Uint, cosmos.Uint, error) {
	if stakeRune.Add(poolRune).IsZero() {
		return cosmos.ZeroUint(), cosmos.ZeroUint(), errors.New("total RUNE in the pool is zero")
	}
	if stakeAsset.Add(poolAsset).IsZero() {
		return cosmos.ZeroUint(), cosmos.ZeroUint(), errors.New("total asset in the pool is zero")
	}
	if poolRune.IsZero() || poolAsset.IsZero() {
		return stakeRune, stakeRune, nil
	}
	P := cosmos.NewDecFromBigInt(oldPoolUnits.BigInt())
	R := cosmos.NewDecFromBigInt(poolRune.BigInt())
	A := cosmos.NewDecFromBigInt(poolAsset.BigInt())
	r := cosmos.NewDecFromBigInt(stakeRune.BigInt())
	a := cosmos.NewDecFromBigInt(stakeAsset.BigInt())

	// (2 r + R) (a + A)
	slipAdjDenominator := (r.MulInt64(2).Add(R)).Mul(a.Add(A))
	// ABS((R a - r A)/((2 r + R) (a + A)))
	var slipAdjustment cosmos.Dec
	if R.Mul(a).GT(r.Mul(A)) {
		slipAdjustment = R.Mul(a).Sub(r.Mul(A)).Quo(slipAdjDenominator)
	} else {
		slipAdjustment = r.Mul(A).Sub(R.Mul(a)).Quo(slipAdjDenominator)
	}
	// (1 - ABS((R a - r A)/((2 r + R) (a + A))))
	slipAdjustment = cosmos.NewDec(1).Sub(slipAdjustment)

	// ((P (a R + A r))
	numerator := P.Mul(a.Mul(R).Add(A.Mul(r)))
	// 2AR
	denominator := cosmos.NewDec(2).Mul(A).Mul(R)
	stakeUnits := numerator.Quo(denominator).Mul(slipAdjustment)
	newPoolUnit := P.Add(stakeUnits)

	return cosmos.NewUintFromBigInt(newPoolUnit.TruncateInt().BigInt()), cosmos.NewUintFromBigInt(stakeUnits.TruncateInt().BigInt()), nil
}
```

@musnit I think if the pool ownership calculations accounted for slip it would clear up the vulnerability you're asking about. Is that correct?

Reply:
Asana task for these changes here: https://app.asana.com/0/1196711589172801/1198984611635837/f